### PR TITLE
[export] hook up mark_dynamic to export Dims

### PR DESCRIPTION
**Pull request by pianpwk**
Adds Dim.DYNAMIC which calls torch._dynamo.mark_dynamic() in the backend. Similar to Dim.AUTO in that it does automatic inference for ranges & relations, but errors out for specializations.

### CODE REVIEW

File: test/export/test_export.py [hunk 1]
```diff
@@ -7757,6 +7757,82 @@ def test_dynamic_shapes_serdes_user_errors(self):
...
+
+        # dynamic should complain when force specialized
+        class Specialize(torch.nn.Module):
+            def forward(self, x):
+                torch._check(x.shape[0] == 4)
+                return x + 2
+
+        with self.assertRaisesRegex(
+            torch._dynamo.exc.UserError,
+            r"Not all values of RelaxedUnspecConstraint.* are valid because .* was inferred to be a constant",
```
Comment (avikchaudhuri) [inline test/export/test_export.py:7805]: nit: we shouldn't be exposing internal names like `RelaxedUnspecConstraint` in user-facing errors, they're noise.

**Review by avikchaudhuri — APPROVED** [review]
